pkg/daemon/update.go

```diff
@@ -1252,23 +1258,40 @@ func (dn *Daemon) writeUnits(units []ign3types.Unit) error {
 		}
 
 		// if the unit doesn't note if it should be enabled or disabled then
-		// skip all linking.
-		// if the unit should be enabled, then enable it.
-		// otherwise the unit is disabled. run disableUnit to ensure the unit is
-		// disabled. even if the unit wasn't previously enabled the result will
-		// be fine as disableUnit is idempotent.
+		// honour system presets. This to account for an edge case where you
+		// deleted a MachineConfig that enabled/disabled the unit to revert,
+		// but the unit itself is referenced in other MCs. deleteStaleData() will
+		// catch fully deleted units.
+		// if the unit should be enabled/disabled, then enable/disable it.
+		// this is a no-op if the system wasn't change this iteration
+		// Also, enable and disable as one command, as if any operation fails
+		// we'd bubble up the error anyways, and we save a lot of time doing this.
+		// Presets must be done individually as we don't consider a failed preset
+		// as an error, but it would cause other presets that would have succeeded
+		// to not go through.
+
 		if u.Enabled != nil {
 			if *u.Enabled {
-				if err := dn.enableUnit(u); err != nil {
-					return err
-				}
-				glog.V(2).Infof("Enabled systemd unit %q", u.Name)
+				enabledUnits = append(enabledUnits, u.Name)
```

Comment:
I think it'd probably be better to do a "diff" here - only enable/disable units that changed between the two configs.

Otherwise I worry that this may cause unexpected side effects.

Reply:
So this doesn't change any existing behaviour. Today we re-write all enable/disables anyways (via the hardlink to multi-user.target.wants) just like how we re-write any file/unit again. I'm not sure if we want to make the exception

Reply:
Right, but I suspect part of the problem here is that going via `systemctl` may touch the DBus API doing a bunch of synchronous requests; that's going to slow things down.

Another angle is to try `SYSTEMD_OFFLINE=1` [in the environment](https://github.com/systemd/systemd/blob/master/docs/ENVIRONMENT.md)?  

Of course if we implemented https://github.com/openshift/machine-config-operator/issues/1190 this issue would also go away because we'd only be enabling in the new root, so shouldn't be talking to the running systemd either.